### PR TITLE
[Debt] Refactor authorizedtoview scopes

### LINKED ISSUE #11041: ♻️ Rewrite authorizedToView scopes so they fall-through to least permissive option
## ♻️ Debt/Refactor

Our authorizedToView scopes are written to add conditions to a query, and then return the query at the end. However, this means that if all of our conditionals are negative, the result is the least permissive query possible. It would be more difficult to make a mistake positive conditionals are required to return a non-empty query.

## 🕵️ Details

This is what we did for Policy methods a while back. Hopefully, it should make the logic flow of these scopes easier to understand and more difficult to mess up.

## 🙋‍♀️ Proposed Solution

Essentially, the final line of an authorizedToView query should be `return $query->where('id', null);`, which will force the query to return nothing. This in turn forces us to check for all relevant permissions and return early if the user has them. If we forget to check a permission, the scope will filter out more results, not fewer.

## ✅ Acceptance Criteria

A set of assumptions which, when tested, verify that the debt was addressed and expected functionality has not been affected.

- [ ] The final line of any `authorizedToView`-type scope is `return $query->where('id', null);`
- [ ] An `authorizedToView`-type scope returns all the items for which the matching policy method would return true.

### PR DESCRIPTION
**Pull request by petertgiles**
🤖 Resolves #11041

## 👋 Introduction

This branch refactors the AuthorizedToAction scopes to make them easier to maintain and safer.  It also adds tests to help document expected behaviour and guard against regression bugs.

## 🕵️ Details

The general strategy used here:
- Test for the most privileged conditions and return early if true
- Fall through to progressively less privileged conditions
- Don't test for permissions twice (hasSomePermission)
- [Type hint a void return](https://laravel.com/docs/10.x/eloquent#local-scopes) - Laravel will not use a different query if you return one

I also added some extra logging the privileged user checker.

I also split User::scopeAuthorizedToViewAsAdmin to (hopefully) make it a little clearer and easier to test.

I'm not sure which roles should be tested right now since we're in the middle of the communities feature.  I wrote tests for the "legacy" roles.  Hopefully they can be a good model for extending when the new roles become active.

I noticed two discrepancies from the role specs:
- Platform admins does not have `view-any-submittedApplication` in the spec but does in the seeder
- Request responder does not have `view-any-user` in the spec but does in the seeder

## 🧪 Testing

1. Rebuild app - no permissions changed so no need to run rolePermissionSeder.

- [ ] App behaves normally (especially access to notifications, pools, candidates, and users)
- [ ] Tests pass
- [ ] Assertions in new tests are correct

### CODE REVIEW

File: api/app/Checkers/ProtectedRequestUserChecker.php [hunk 1]
```diff
@@ -56,6 +57,8 @@ public function currentUserHasRole(
         bool $requireAll = false
     ): bool {
         if (! $this->isSafeToUseRole($name)) {
+            Log::warning('Tried to unsafely use role '.json_encode($name));
```
Comment (petertgiles) [inline api/app/Checkers/ProtectedRequestUserChecker.php:60]: This is not directly related to the AC - just that it would have saved me lots of time if it existed before.  :face_exhaling:

File: api/app/Models/Pool.php [hunk 2]
```diff
@@ -541,43 +541,47 @@ public static function scopeClassifications(Builder $query, ?array $classificati
...
     /**
-     * Filter for pools the user is allowed to view admin information for, based around assessment plan permissions
+     * Filter for pools the user is allowed to admin, based on scopeAuthorizedToAdmin
      */
-    public static function scopeAuthorizedToViewAsAdmin(Builder $query, ?bool $canAdmin): Builder
+    public static function scopeCanAdmin(Builder $query, ?bool $canAdmin): Builder
     {
-        if (empty($canAdmin)) {
-            return $query;
+        if (! empty($canAdmin)) {
```
Comment (petertgiles) [inline api/app/Models/Pool.php:548]: This maintains the existing behaviour - but maybe it should be `if ($canAdmin)` so it only runs if truthy?

File: api/tests/Unit/PoolCandidateAuthorizationScopeTest.php [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+                User::factory()
+                    ->asCommunityManager()
+                    ->create()
+            );
+
+        $candidateIds = PoolCandidate::authorizedToView()->get()->pluck('id');
+        assertEqualsCanonicalizing([], $candidateIds->toArray());
+    }
+
+    // a platform admin should be able to view submitted candidates in any team
```
Comment (petertgiles) [inline api/tests/Unit/PoolCandidateAuthorizationScopeTest.php:169]: Technically, this goes against the spec.

File: api/tests/Unit/UserAuthorizationScopeTest.php [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+
+        // can also see the user with an application submitted to team 1's pool
+        $userIds = User::authorizedToView()->get()->pluck('id');
+        assertEqualsCanonicalizing([
+            $poolOperator->id,
+            $this->user1->id,
+        ], $userIds->toArray());
+    }
+
+    // a request responder should be able to view any user
```
Comment (petertgiles) [inline api/tests/Unit/UserAuthorizationScopeTest.php:128]: Technically, this goes against the spec.

**Review by petertgiles — COMMENTED** [review]
Some notes.

**Comment by tristan-orourke** [comment]
You're supposed to return void from scopes? Huh. And you can still chain them with other query builder methods?

**Comment by petertgiles** [comment]
> You're supposed to return void from scopes? Huh. And you can still chain them with other query builder methods?

You _can_ return another query builder but I found out after much head-scratching that Lighthouse will not use it and wil stick with the original.  So probably safer not to do it at all.

**Comment by esizer** [comment]
> > You're supposed to return void from scopes? Huh. And you can still chain them with other query builder methods?
> 
> You _can_ return another query builder but I found out after much head-scratching that Lighthouse will not use it and will stick with the original. So probably safer not to do it at all.

This is really good info to have :sweat_smile: Thanks for finding that out :+1:

File: api/app/Models/Pool.php [hunk 5]
```diff
@@ -632,43 +636,45 @@ public function scopeOrderByPoolBookmarks(Builder $query, ?array $args): Builder
...
+            }
+        });
+        $filterCountAfter = count($query->getQuery()->wheres);
+        if ($filterCountAfter > $filterCountBefore) {
+            return;
         }
 
-        return $query;
+        // fall through - anyone can view a published pool
+        $query->where('published_at', '<=', Carbon::now()->toDateTimeString());
```
Comment (tristan-orourke) [inline api/app/Models/Pool.php:677]: This is true, but it's already capturedby `view-any-publishedPool`, which is assigned to the baseUser role.

File: api/app/Models/Pool.php [hunk 6]
```diff
@@ -632,43 +636,45 @@ public function scopeOrderByPoolBookmarks(Builder $query, ?array $args): Builder
...
+            }
 
-                return $query;
-            });
+            if ($user?->isAbleTo('view-any-publishedPool')) {
+                $query->orWhere('published_at', '<=', Carbon::now()->toDateTimeString());
+            }
+        });
+        $filterCountAfter = count($query->getQuery()->wheres);
+        if ($filterCountAfter > $filterCountBefore) {
```
Comment (tristan-orourke) [inline api/app/Models/Pool.php:672]: Hmm. Won't this always be true, because of the `where` at line 651?

File: api/app/Models/User.php [hunk 7]
```diff
@@ -1041,44 +1039,60 @@ public function scopeAuthorizedToViewSpecific(Builder $query)
...
+            return;
+        }
 
-        return $query;
+        // fall through - query will return nothing
+        $query->where('id', null);
     }
 
-    public function scopeAuthorizedToView(Builder $query)
+    public function scopeAuthorizedToView(Builder $query): void
```
Comment (tristan-orourke) [inline api/app/Models/User.php:1056]: I don't remember why we separated this from `scopeAuthorizedToViewSpecific` and I don't see any advantage. Do you? 🤔

File: api/tests/Unit/PoolAuthorizationScopeTest.php [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+            $this->poolPublished1->id,
+            $this->poolClosed1->id,
+            $this->poolArchived1->id,
+            $this->poolDraft2->id,
+            $this->poolPublished2->id,
+            $this->poolClosed2->id,
+            $this->poolArchived2->id,
+        ], $poolIds->toArray());
+    }
+}
```
Comment (tristan-orourke) [inline api/tests/Unit/PoolAuthorizationScopeTest.php:289]: Could you add a comment: `TODO: add tests for new roles as part of #10609`

File: api/tests/Unit/PoolCandidateAuthorizationScopeTest.php (hunk 3 above)
Comment (tristan-orourke) [reply api/tests/Unit/PoolCandidateAuthorizationScopeTest.php:169]: Thanks, I've updated the spec in this case.

File: api/tests/Unit/PoolCandidateAuthorizationScopeTest.php [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+            );
+
+        // there is one submitted candidate in each team's pool
+        $candidateIds = PoolCandidate::authorizedToView()->get()->pluck('id');
+        assertEqualsCanonicalizing([
+            $this->candidateSubmitted1B->id,
+            $this->candidateSubmitted2A->id,
+        ], $candidateIds->toArray());
+    }
+}
```
Comment (tristan-orourke) [inline api/tests/Unit/PoolCandidateAuthorizationScopeTest.php:186]: Could you add the reminder here as well? `TODO: add tests for new roles as part of #10609`

File: api/tests/Unit/UserAuthorizationScopeTest.php [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+        $this->platformAdmin = User::factory()
+            ->asAdmin()
+            ->create();
+
+        $this->pool1 = Pool::factory()
+            ->for($this->platformAdmin)
+            ->published()
+            ->afterCreating(function (Pool $p) {
+                // normally this happens only when the process operator is added - might want to reconsider that logic
+                $p->team()->create(['name' => 'pool-'.$p->id]);
```
Comment (tristan-orourke) [inline api/tests/Unit/UserAuthorizationScopeTest.php:52]: If you want to use this for the PoolOperator role, you actually want to use `$p->legacyTeam()`. The `team` relation going forward is used for the new ProcessOperator role. (This will be much less confusing after #10368.) Setting team_id, like you did in api/tests/Unit/PoolCandidateAuthorizationScopeTest.php, still works.

File: api/tests/Unit/UserAuthorizationScopeTest.php [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+        assertEqualsCanonicalizing([
+            $this->user1->id,
+        ], $userIds->toArray());
+    }
+
+    // an pool operator should be able to view themselves and any users with submitted candidates in their pools
+    public function testViewAsPoolOperator(): void
+    {
+        $poolOperator = User::factory()
+            ->asPoolOperator($this->pool1->team->name)
```
Comment (tristan-orourke) [inline api/tests/Unit/UserAuthorizationScopeTest.php:115]:
```suggestion
            ->asPoolOperator($this->pool1->legacyTeam->name)
```

File: api/tests/Unit/UserAuthorizationScopeTest.php [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+        Auth::shouldReceive('user')
+            ->andReturn($this->platformAdmin);
+
+        $userIds = User::authorizedToViewBasicInfo()->get()->pluck('id');
+        assertEqualsCanonicalizing([
+            $this->platformAdmin->id,
+            $this->user1->id,
+            $this->user2->id,
+        ], $userIds->toArray());
+    }
```
Comment (tristan-orourke) [inline api/tests/Unit/UserAuthorizationScopeTest.php:262]: Could you add the reminder here as well? `TODO: add tests for new roles as part of #10609`

File: api/app/Models/Pool.php (hunk 6 above)
Comment (tristan-orourke) [reply api/app/Models/Pool.php:672]: I guess not! I tested it in tinker. And `->where( function ($q) { ... } )` with an empty inner function doesn't actually increase the where count. Could you add a little comment mentioning that this has been considered and isn't a problem?

**Review by tristan-orourke — CHANGES_REQUESTED** [review]
Looks good. The `filterCountAfter` check is a smart approach which avoids having to check permissions twice. I just have some minor questions and a note for a test.

File: api/tests/Unit/UserAuthorizationScopeTest.php (hunk 4 above)
Comment (tristan-orourke) [reply api/tests/Unit/UserAuthorizationScopeTest.php:128]: This one I'll scope out into another ticket.

File: api/app/Models/Pool.php (hunk 2 above)
Comment (petertgiles) [reply api/app/Models/Pool.php]: Oh, interesting!  I agree that if it confused us both it should change.
[clarify scopeCanAdmin test](https://github.com/GCTC-NTGC/gc-digital-talent/pull/11138/commits/c81a8bc0e1d77913b9a145617ef493bf640dc555)

File: api/tests/Unit/UserAuthorizationScopeTest.php (hunk 12 above)
Comment (petertgiles) [reply api/tests/Unit/UserAuthorizationScopeTest.php:262]: [extra comments](https://github.com/GCTC-NTGC/gc-digital-talent/pull/11138/commits/dbda192c713a308da0628b1501ca454a706f879a)

File: api/tests/Unit/PoolCandidateAuthorizationScopeTest.php (hunk 9 above)
Comment (petertgiles) [reply api/tests/Unit/PoolCandidateAuthorizationScopeTest.php:188]: [extra comments](https://github.com/GCTC-NTGC/gc-digital-talent/pull/11138/commits/dbda192c713a308da0628b1501ca454a706f879a)

File: api/tests/Unit/PoolAuthorizationScopeTest.php (hunk 8 above)
Comment (petertgiles) [reply api/tests/Unit/PoolAuthorizationScopeTest.php:291]: [extra comments](https://github.com/GCTC-NTGC/gc-digital-talent/pull/11138/commits/dbda192c713a308da0628b1501ca454a706f879a)

File: api/app/Models/Pool.php [hunk 13]
```diff
@@ -554,43 +554,47 @@ public static function scopeClassifications(Builder $query, ?array $classificati
...
-                    $query->orWhereHas('legacyTeam', function (Builder $query) use ($teamIds) {
-                        return $query->whereIn('id', $teamIds);
-                    });
-                } else {
-                    return $query->where('id', null); // when the user can't see any assessment plans
                 }
-            }
-            );
+
+                $query->orWhereHas('legacyTeam', function (Builder $query) use ($teamIds) {
```
[the remark this answers is not in the document]
Comment (petertgiles) [reply api/app/Models/Pool.php:589]: [extra comments](https://github.com/GCTC-NTGC/gc-digital-talent/pull/11138/commits/dbda192c713a308da0628b1501ca454a706f879a)

File: api/app/Models/Pool.php (hunk 5 above)
Comment (petertgiles) [reply api/app/Models/Pool.php:678]: Even guests have `view-any-publishedPool`.  But how can you call `isAbleTo` on a null user object?

File: api/app/Models/Pool.php (hunk 6 above)
Comment (petertgiles) [reply api/app/Models/Pool.php:673]: [extra comments](https://github.com/GCTC-NTGC/gc-digital-talent/pull/11138/commits/dbda192c713a308da0628b1501ca454a706f879a)

File: api/app/Models/User.php (hunk 7 above)
Comment (petertgiles) [reply api/app/Models/User.php:1056]: It is an odd pattern.  I think it's just that we factored out the bulk of the logic and one scope has a quick return for the "any-basicinfo" permission.  Maybe we could find a better name for `scopeAuthorizedToViewSpecific`?
image

File: api/app/Models/Pool.php (hunk 5 above)
Comment (tristan-orourke) [reply api/app/Models/Pool.php:678]: Nope, that's a great point! This is necessary to accommodate guests.

File: api/tests/Unit/UserAuthorizationScopeTest.php (hunk 11 above)
Comment (petertgiles) [reply api/tests/Unit/UserAuthorizationScopeTest.php]: [fix user legacy team check](https://github.com/GCTC-NTGC/gc-digital-talent/pull/11138/commits/aebb9a4a9f36ba5e51c932ba1f8977a1713dd48d)

File: api/app/Models/User.php (hunk 7 above)
Comment (tristan-orourke) [reply api/app/Models/User.php:1062]: I see. The difference is that `authorizedToViewSpecific` skips checking `view-any-user` permissions, allowing it to be called by both `authorizedToView` and `authorizedToViewBasicInfo` without any duplication.

However... this still seems pointless and confusing to me now. I suggest ditching `authorizedToViewSpecific`, moving that logic into `authorizedToView`, and just calling `authorizedToView` from `authorizedToViewBasicInfo`. Any objection?

File: api/tests/Unit/UserAuthorizationScopeTest.php (hunk 10 above)
Comment (tristan-orourke) [reply api/tests/Unit/UserAuthorizationScopeTest.php]: Good catch!

**Review by tristan-orourke — COMMENTED** [review]
Thanks, you've addressed pretty much everything, I'd just push more that refactoring of the user authorizedToViewSpecific scope before merging.

File: api/app/Models/User.php (hunk 7 above)
Comment (petertgiles) [reply api/app/Models/User.php]: Yeah, good call.
[refactor scopeAuthorizedToViewSpecific](https://github.com/GCTC-NTGC/gc-digital-talent/pull/11138/commits/0f123831f9eea3959af35a3859d67b88cbc02f6e)

**Review by tristan-orourke — APPROVED** [review]
